extensions/amp-viewer-integration/0.1/messaging/messaging.js

```diff
@@ -79,7 +82,11 @@ export class WindowPortEmulator {
    */
   addEventListener(eventType, handler) {
     this.win_.addEventListener('message', (event) => {
-      if (event.origin == this.origin_ && event.source == this.target_) {
+      if (
+        (event.data.messagingToken === this.token_ ||
+          event.origin == this.origin_) &&
```

Comment:
Just a small suggestion: if I understand correctly, this means having a messaging token completely disables the origin checking behaviour.

Would it be possible to keep both (i.e. change `||` to `&&` here)? Then you can just use `'null'` as the origin parameter when you create a `WindowPortEmulator` when there's no origin (this is what we do for the AMP for Email viewer).

Reply:
I actually tried passing `'null'` as the origin parameter https://github.com/ampproject/amphtml/issues/28225#issuecomment-625461308, but the problem is that when the `WindowPortEmulator` receives a message from the viewer's parent window, `event.origin` actually has the `origin` of the parent window, making this check fail.

Reply:
token_ is null if it was not specified, so a message with messagingToken=null could bypass the origin check, if the token was not expected.

Btw, other token checking is done in the Messaging class, not sure how this interacts with that, would be better to have all this logic in a single class. Also, the Messaging class implemented this check in a better way here without the flaw mentioned above: https://github.com/ampproject/amphtml/blob/b65705aae8892fe3c4dceab624296fbfbee58315/extensions/amp-viewer-integration/0.1/messaging/messaging.js#L301